### PR TITLE
Load NFT off-chain metadata on backend

### PR DESCRIPTION
**Pull request by ytx1991**
1. Off-chain data will be fetched asynchronously when the wallet manager discovered a NFT
2. The default cache folder is .chia/mainet/nft_cache . User can config the path by setting the "nft_metadata_cache_path"
3. It will check if there is a valid cached metadata before fetching remote metadata.
4. User can delete or modify the cached metadata. They will be fetched again when calling nft_get_nfts.
5. The cached metadata will be removed when user lost the NFT ownership.
6. Add pagination for the nft_get_nfts. Use "start_index" and "num" to control. Default value is returning all NFTs.
7. Added a new string field "off_chain_data" into the response. For saving backend resource, off-chain data will not be serialized. GUI needs to serialize it.

### CODE REVIEW

**Comment by ytx1991** [comment]
Not sure if this should merge to 1.5.1

**Review by altendky — DISMISSED** [review]
What about `nft_off_chain.py` requires that it be added to the mypy leniency list?

**Comment by wallentx** [comment]
> Not sure if this should merge to 1.5.1

Since this will be for 1.6.0, I think targeting this branch is okay, then switch to the next release branch after we merge this. It might be good to convert this PR to a draft for now

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (trepca) [inline chia/wallet/nft_wallet/nft_off_chain.py:26]: please use `aiohttp` so it doesn't block (await)

File: tests/wallet/nft_wallet/test_nft_off_chain.py
Comment (trepca) [inline tests/wallet/nft_wallet/test_nft_off_chain.py:69]: there are no asserts in this function, why is it prefixed with `test_`? looks like a helper function, so makes more sense to name it `create_test_full_puzzle`

File: chia/wallet/nft_wallet/nft_wallet.py
Comment (trepca) [inline chia/wallet/nft_wallet/nft_wallet.py:265]: if user receives a lot of new coins this will fork same amount of processes and given fetching metadata is mainly I/O bound an `asyncio.create_task` should be sufficient?
another potential issue here is that if we receive 1000 NFTs, we would issue 1000 requests crushing users bandwidth/connection, so should be rate limited

Comment (altendky) [reply chia/wallet/nft_wallet/nft_wallet.py:265]: I would rather not spawn lots of tasks then block them all on a semaphore or such so they sit around doing nothing.  But, that's what we do now and I haven't gotten my reference async worker pool into use so go for it for right now.  The idea being to spawn as many workers as you want working and have them just read from a queue of work to be done.

File: chia/rpc/wallet_rpc_api.py
Comment (paninaro) [inline chia/rpc/wallet_rpc_api.py:1511]: I appreciate building out support for pagination. One GUI use case that this RPC doesn't natively handle is fetching _all_ NFTs across all wallets. At the moment, the GUI is calling this RPC for each NFT wallet and aggregating the results into a single list.

I can see this pagination implementation being useful when selecting a specific wallet in the GUI/CLI, but if "All NFTs" is selected, this form of pagination is not going to work.

Would it be too much trouble to add a mode here to fetch all NFTs? Perhaps we could just make the `"wallet_id"` param optional -- if None, fetch all NFTs.

Comment (paninaro) [reply chia/rpc/wallet_rpc_api.py:1511]: I think it should be as simple as adding an outer loop that iterates over NFT wallets

File: chia/rpc/wallet_rpc_api.py
Comment (paninaro) [inline chia/rpc/wallet_rpc_api.py:1518]: nit: Maybe just save this value to a local variable?

Comment (paninaro) [reply chia/rpc/wallet_rpc_api.py:1518]: I see the same line repeated a few times throughout this file

File: chia/wallet/nft_wallet/nft_info.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_info.py:78]: Any objection to naming this `off_chain_metadata`? Just in case we want to support exposing other types of off-chain data at some point.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:5]: @ytx1991 @altendky I was reading about https://www.python-httpx.org/ recently -- a drop-in replacement for `requests`, but with async support (using a different API). I don't see other uses of requests in the codebase. Is it worth considering this other package?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:17]: @altendky is this pattern ok? Creating a directory on module import?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:26]: We should enforce some size limit (maybe around 1 MB) to ensure that abusive/airdropped NFTs don't fill up the cache. If the metadata is larger than that, I think it's fine for the client to support manually fetching the metadata.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:45]: Debug logging for failed metadata hash verification?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:47]: Log the exception? If the file exists and we fail to read it, it might be useful info.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:37]: Minor suggestion: create a function to get the cache filename for an NFT

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [reply chia/wallet/nft_wallet/nft_off_chain.py:17]: This probably isn't necessary since the directory is created in `write_off_chain_metadata()`, right?

File: chia/rpc/wallet_rpc_api.py
Comment (paninaro) [inline chia/rpc/wallet_rpc_api.py:1518]: I think including/fetching the off-chain metadata should be optional. Can we use a request param `"include_off_chain_metadata"` to control that?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:77]: I'm concerned that get_off_chain_metadata is called in several locations, without the ability to disable the synchronous fetching of the data (if not already cached). For some operations I think we'll want that control (`nft_set_nft_did` for example doesn't need to fetch the data).

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [inline chia/wallet/nft_wallet/nft_off_chain.py:15]: Shouldn't we also update initial-config.yaml to include this with a default value? Otherwise users will have a hard time figuring out how to override the default path (without looking at this code).

File: chia/wallet/nft_wallet/nft_wallet.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_wallet.py:265]: Are you referring thread pooling?

Comment (altendky) [reply chia/wallet/nft_wallet/nft_wallet.py:265]: No, asyncio tasks that just process work from a queue.  But sure, same idea as a pool of threads or processes.  I have an old PR I need to resurrect to give us a reusable form for this.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [reply chia/wallet/nft_wallet/nft_off_chain.py:5]: For now we would just use the aiohttp client wouldn't we?  But sure, my impression, without being a major http user until here, is that httpx is a nice and newer lib.  There are options other than aiohttp for serving http as well that we could consider.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [reply chia/wallet/nft_wallet/nft_off_chain.py:17]: I don't like it, no.  Importing should not have side effects.

Separately from what we do here probably, there are proper locations for caches vs. tmp files vs. program data vs. config...  https://pypi.org/project/appdirs/ would be relevant if at some point we decided to follow the standards for that.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [reply chia/wallet/nft_wallet/nft_off_chain.py:47]: What are the possible exceptions here?  We certainly don't want to consume `NameError` or such.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:41]: Don't bother pre-checking.  You will still need to sensibly handle the exception since there's a race condition here where the file could get deleted between the existence check and the read attempt.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_off_chain.py:77]: For most cases we don't need to fetch metadata on the fly for any API. The metadata should already be cached at this point of time, unless user deleted or changed them manually.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:100]: I see that `cache_path` being a string is propagated from existing code so I won't request a change on it here.  But, be aware that such deserialization should be done at the boundaries rather than pushing the serialized form all throughout the code.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:104]: What's wrong with the case where `cache_path` is `""`?

**Review by altendky — COMMENTED** [review]
It looks like at least a couple conversations, I didn't check more, were resolved without making the changes.  Is there an incoming commit with code responses to those?

File: chia/wallet/nft_wallet/nft_wallet.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_wallet.py]: I think we don't have to create another thread for loading metadata. This thread is only for this coin right?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_off_chain.py:104]: The default value is "". Can we set None in yaml?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_off_chain.py:17]: If there are all configs we have to pass many parameter to the methods. Can you point me the helper function you mentioned?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (trepca) [inline chia/wallet/nft_wallet/nft_off_chain.py:31]: wouldn't this load all fetched file content into memory?

Comment (trepca) [reply chia/wallet/nft_wallet/nft_off_chain.py:31]: let's say metadata file is 100GB in size, wouldn't this try to load the whole file content into `text` variable?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (trepca) [inline chia/wallet/nft_wallet/nft_off_chain.py:31]: timeout by default is 5 minutes, which has a big impact to how long it could take to load all metadata (especially over time when NFTs will have a lot of obsolete metadata URLs). 30 seconds should be enough.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (trepca) [inline chia/wallet/nft_wallet/nft_off_chain.py:79]: I think if `fetch_off_chain_metadata` always returns None due to bad connection/not existing anymore, we still try to fetch it again forever whenever we call `get_nft_info_from_puzzle`, right?

**Review by trepca — CHANGES_REQUESTED** [review]

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (trepca) [reply chia/wallet/nft_wallet/nft_off_chain.py]:
```
>>> yaml.safe_load("var: null")
{'var': None}
>>> yaml.safe_load("var: ")
{'var': None}
>>> yaml.safe_load("var: ''")
{'var': ''}
>>> yaml.safe_load('var: ""')
{'var': ''}
```

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_off_chain.py:33]: What is the solution for this in Python? I was thinking use content-length to check the file size but it is not reliable.

File: chia/wallet/nft_wallet/nft_wallet.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_wallet.py]: I tried asyncio, but it doesn't work here because of it is in an async event loop.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_off_chain.py:52]: Why we don't want to catch NameError? It will reload metadata no matter what kind of exception.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (ytx1991) [reply chia/wallet/nft_wallet/nft_off_chain.py:79]: Yes, if there isn't a valid cached file it will keep fetching until success.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (trepca) [reply chia/wallet/nft_wallet/nft_off_chain.py:34]: you can use the streaming API: https://docs.aiohttp.org/en/stable/streams.html

for example:
```python
async with aiohttp.ClientSession() as session:
    async with session.get(str(uri, "utf-8")) as response:
        if response.status == 200:
            async for chunk in response.content.iter_chunked(1024):
                file_obj.write(chunk)
```

**Comment by paninaro** [comment]
Since the normal mode of operation for this is to fetch the metadata when an NFT coin is added, how are we planning to handle upgrades for users that already have their wallets synced?

**Comment by ytx1991** [comment]
> Since the normal mode of operation for this is to fetch the metadata when an NFT coin is added, how are we planning to handle upgrades for users that already have their wallets synced?

I don't know what is the concern. For synced users, the cache will be fetched on the first time of calling nft_get_nfts.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [reply chia/wallet/nft_wallet/nft_off_chain.py:58]: I'm happy with the logging addition. I'm not sure I understand what the `NameError` discussion is about @altendky

**Comment by paninaro** [comment]
> > Since the normal mode of operation for this is to fetch the metadata when an NFT coin is added, how are we planning to handle upgrades for users that already have their wallets synced?
> 
> I don't know what is the concern. For synced users, the cache will be fetched on the first time of calling nft_get_nfts.

My concern is that users that currently have a large collection will have a poor experience immediately after upgrading. The GUI may appear to be unresponsive while the metadata is being fetched. I don't want to hold this PR up, but maybe there's something we can do as a follow-up to make this experience better.

**Review by paninaro — DISMISSED** [review]
lgtm

File: chia/rpc/wallet_rpc_api.py
Comment (trepca) [inline chia/rpc/wallet_rpc_api.py:1515]: if you're adding pagination, there should also be pagination tests

**Review by trepca — CHANGES_REQUESTED** [review]
thanks, lets add tests for pagination and should be it

**Review by trepca — DISMISSED** [review]

**Review by paninaro — DISMISSED** [review]
lgtm

File: chia/util/initial-config.yaml
Comment (altendky) [inline chia/util/initial-config.yaml:495]: See the other comment about `chia.util.path.path_from_root()`.

```suggestion
  nft_metadata_cache_path: "nft_cache"
```

File: setup.py
Comment (altendky) [inline setup.py:36]: We already have this above.

```suggestion
```

File: chia/rpc/wallet_rpc_api.py
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py:1533]: Doesn't the RPC framework handle logging and returning errors?  If kept more as it already is, it needs to be made to not consume `asyncio.CancelledError`.

```suggestion
        wallet_id = request.get("wallet_id", None)
        nfts = []
        if wallet_id is not None:
            nft_wallet: NFTWallet = self.service.wallet_state_manager.wallets[wallet_id]
            nfts.extend(nft_wallet.get_current_nfts())
        else:
            for wallet in self.service.wallet_state_manager.wallets.values():
                if wallet.type() == WalletType.NFT.value:
                    nfts.extend(wallet.get_current_nfts())
        start_index = request.get("start_index", 0)
        num = request.get("num", len(nfts))
        nft_info_list = []
        count = 0
        for nft in nfts:
            if count >= start_index and count - start_index < num:
                nft_info = await nft_puzzles.get_nft_info_from_puzzle(
                    nft,
                    self.service.wallet_state_manager.config,
                    request.get("include_off_chain_metadata", False),
                    request.get("ignore_size_limit", False),
                )
                nft_info_list.append(nft_info)
            count += 1
        return {"wallet_id": wallet_id, "success": True, "nft_list": nft_info_list}
```

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:15]: See the other comment about `chia.util.path.path_from_root()`.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:63]: Don't pre-check, just handle the exception raised in the case it isn't there.  As is, the file could still be deleted between this check and the `.unlink()` below so the exception could still happen.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:118]: Take a look at uses of `chia.util.path.path_from_root()`.  Then remove the `./` from other spots.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:114]: Shouldn't these already be of these types anyways?

```suggestion
        cache_path: str = config.get(CACHE_PATH_KEY, DEFAULT_CACHE_PATH)
        hash_length: int = config.get(PREFIX_HASH_LENGTH_KEY, DEFAULT_PREFIX_HASH_LENGTH)
```

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [reply chia/wallet/nft_wallet/nft_off_chain.py:58]: You don't want to catch `NameError` because it indicates a variable name typo or such.  That should fail out hard so we see it quickly and fix it, not hide in the logs where we forget to look while developing and only hear about it when a user sends us a log asking about it.  This is a general commentary on us catching all exceptions in lots of places without thinking about, or often even being aware of, all the relevant exceptions.  `NameError` in particular is just an example.

**Review by altendky — DISMISSED** [review]

**Comment by emlowe** [comment]
This is out of 1.6.0, marking as Draft to avoid any accidental merging and removed 1.6.0 label

File: chia/util/initial-config.yaml
Comment (ytx1991) [reply chia/util/initial-config.yaml:495]: ./ is used to distinguish the relative path and absolute path. User may not want to cache the off-chain metadata on their system disk.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (paninaro) [reply chia/wallet/nft_wallet/nft_off_chain.py:60]: What if we just re-raise the exception here after logging it?

**Review by paninaro — APPROVED** [review]
lgtm

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (trepca) [reply chia/wallet/nft_wallet/nft_off_chain.py:60]: have opened a new issue re catching generic exception

File: chia/util/initial-config.yaml
Comment (altendky) [reply chia/util/initial-config.yaml]: This is no different than our other big stuff.  It need not be defaulted differently here.  Using `./` just means it will go any number of places depending what directory the user is in when they run chia.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:121]: Why does this need to deviate from existing usage?

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [inline chia/wallet/nft_wallet/nft_off_chain.py:68]: I think `OSError` would be a lot more apt.  That would avoid caching `AttributeError`, `NameError`, etc that we _want_ to see asap.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [reply chia/wallet/nft_wallet/nft_off_chain.py:60]: This in particular was mostly just a discussion.  Reraising is an thing but sometimes we end up double logging then.  Unexpected exceptions should generally bubble really far up and then be reported clearly to developers.  None of the `except Exception:` logging here provides _any_ information about what happened.

File: chia/wallet/nft_wallet/nft_off_chain.py
Comment (altendky) [reply chia/wallet/nft_wallet/nft_off_chain.py:68]: Also report some info from the exception.

**Pull request merged**
